src/compiler/checker.ts

```diff
@@ -3195,24 +3223,33 @@ namespace ts {
             return type;
         }
 
-        function createUnionProperty(unionType: UnionType, name: string): Symbol {
-            let types = unionType.types;
+        function createUnionOrIntersectionProperty(containingType: UnionOrIntersectionType, name: string): Symbol {
+            let types = containingType.types;
             let props: Symbol[];
             for (let current of types) {
                 let type = getApparentType(current);
                 if (type !== unknownType) {
                     let prop = getPropertyOfType(type, name);
-                    if (!prop || getDeclarationFlagsFromSymbol(prop) & (NodeFlags.Private | NodeFlags.Protected)) {
-                        return undefined;
-                    }
-                    if (!props) {
-                        props = [prop];
+                    if (prop && !(getDeclarationFlagsFromSymbol(prop) & (NodeFlags.Private | NodeFlags.Protected))) {

```

Comment:
Correct, as is the case for union types. When a private or protected property with a particular name occurs in only one constituent type, it is pretty clear that the property should be excluded. It is bit less clear what should happen in cases like the following:

```typescript
class C {
    private x: { a: string };
}

interface I {
    x: { b: string };
}

var obj: C & I;
```

Currently `obj` appears to have no property `x`, but you could argue that it should have a property `x` of type `{ private a: string, b: string }` or something like that.